yetanotherco/aligned_layer
# feat(operator): Update SP1 and Risc0 Versions

# Update SP1 and Risc0 verifier versions

closes #1205 

# This PR:
 - Bumps the `risc0` and `sp1` verifier versions of operators. 
 - Provides retro-compatibility for each version. If proof verification for SP1 and Risc0 between the current and old ('_old') verifier versions fails within the operator it will try verifying using the older version of SP1/Risc0 before returning false.
 - Changes the prompt library in the quiz example from `dialoguer` -> `inquire` to accomadate a dependency conflict.

## To Test:
### Verify network bindings works on a local network:
- `make test_risc_zero_go_bindings_macos`
- `make test_sp1_go_bindings_macos`
- run local testnet and confirm you can submit sp1 and risc0 proofs. via
`make batcher_send_risc0_task`
`make batcher_send_sp1_task`
- confirm examples work:
`cd examples/zkquiz && make answer_quix`
`cd examples/validating-public-input && make generate_risc_zero_fibonacci_proof`

### Verify Retro Compatiblility of Operator Verifiers:
To verify operator retro-compatibility we deploy the rest of infrastructure of Aligned and boot the retro-compatible operator. This tests that network retro-compatible operator can verify SP1/Risc0 proofs from both verifier versions.

Note:
Commands suffixed with `_old` indicate the deprecated version of the SP1/Risc0 verifier that is currently running on `testnet`

To test this have two repos of aligned_layer one set to `testnet` and one to `1205-bump-sp1-and-risc0-version`. 

In `testnet` repo:
```
go clean
go mod tidy
make go_deps
make deps
```
- Then you must run anvil, aggregator and batcher:
```
make anvil_start_with_block_time
```
```
make aggregator_start
```
```
make batcher_start_local
```
 
 In `1205-operator-sp1-risc0-update` repo:
 ```
make deps
```
- Then you can run operator
```
make operator_register_and_start
```
- Send proof from the older SP1 and Risc0 versions from the `testnet` repo.
```
make batcher_send_burst_groth16
make batcher_send_sp1_burst
make batcher_send_risc0_burst
```
Observe the `Risc0 proof verification failed.` and `SP1 proof verification failed.` logs are emitted and the operator successfully verifying the proofs from the old SP1/Risc0 verifiers. If you send a proof from the latest version of SP1/Risc0 you should notice its verification fails in the batcher.

- Keep the operator running on the `1205-operator-sp1-risc0-update` repo.
- In the `testnet` repo, kill the batcher. 
- Navigate to the `1205-bump-sp1-and-risc0-version` and start the batcher within the repo.
- Send proof from both the SP1 and Risc0 versions from the `1205-operator-sp1-risc0-update` repo.
```
make batcher_send_sp1_burst
make batcher_send_risc0_burst
```
Observe the operator successfully verifying the proofs from the new SP1/Risc0 verifiers. If you send a proof from the older version of SP1/Risc0 you should notice its verification fails in the batcher.

## Type of change

Please delete options that are not relevant.


## Checklist
  - [Link to the open issue addressing it]()
